rio_tiler/utils.py

```diff
@@ -88,6 +88,116 @@ def _stats(
     )
 
 
+def get_bands_names(
+    indexes: Optional[Sequence[int]] = None,
+    expression: Optional[str] = None,
+    count: Optional[int] = None,
+) -> List[str]:
+    """Define bands names based on expression, indexes or band count."""
+    if expression:
+        return expression.split(",")
+
+    elif indexes:
+        return [str(idx) for idx in indexes]
+
+    elif count:
+        return [str(idx + 1) for idx in range(count)]
+
+    else:
+        raise ValueError(
+            "one of expression or indexes or count must be passed to define band names."
+        )
+
+
+def get_array_statistics(
+    data: numpy.ma.array,
+    categorical: bool = False,
+    categories: Optional[List[float]] = None,
+    percentiles: List[int] = [2, 98],
+    **kwargs: Any,
+) -> List[Dict[Any, Any]]:
+    """Calculate array statistics.
+
+    Args:
+        data (numpy.ndarray): Input array data to get the stats from.
+        categorical (bool): treat input data as categorical data. Defaults to False.
+        categories (list of numbers, optional)
+        percentiles (list of numbers, optional): List of percentile values to calculate. Defaults to `[2, 98]`.
+        kwargs (optional): Options to forward to numpy.histogram function (only applies for non-categorical data).
+
+    Returns:
+        list of dict
+
+    Examples:
+        >>> {
+            'percentiles': [38, 147],
+            'min': 20,
+            'max': 180,
+            'std': 28.123562304138662,
+            'histogram': [
+                [1625, 219241, 28344, 15808, 12325, 10687, 8535, 7348, 4656, 1208],
+                [20.0, 36.0, 52.0, 68.0, 84.0, 100.0, 116.0, 132.0, 148.0, 164.0, 180.0]
+            ],
+            'valid_percent': 0.5
+        }
+
+    """
+    if len(data.shape) < 3:
+        data = numpy.expand_dims(data, axis=0)
+
+    output: List[Dict[Any, Any]] = []
+    percentiles_names = [f"percentile_{int(p)}" for p in percentiles]
+
+    for b in range(data.shape[0]):
+        keys, counts = numpy.unique(data[b].compressed(), return_counts=True)
+
+        valid_pixels = float(numpy.ma.count(data[b]))
+        masked_pixels = float(numpy.ma.count_masked(data[b]))
+        valid_percent = round((valid_pixels / data[b].size) * 100, 2)
+        info_px = {
+            "valid_pixels": valid_pixels,
+            "masked_pixels": masked_pixels,
+            "valid_percent": valid_percent,
+        }
+
+        if categorical:
+            out_dict = dict(zip(keys.tolist(), counts.tolist()))
+            h_keys = (
+                numpy.array(categories).astype(keys.dtype) if categories else keys
+            ).tolist()
+            histogram = [
+                [out_dict[x] for x in h_keys],
+                h_keys,
+            ]
```

Comment:
if the data is set as `categorical` the histogram is then set to be the count of each values.